File: Data Structures & Algorithms/surrounded-regions/submission-15.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        rows = len(board) 
        cols = len(board[0])  
        q = deque()

        for r in range(rows): 
            for c in range(cols): 
                if board[r][c] == 'O' and ( r == 0 or r == rows - 1 
                    or c == 0 or c == cols - 1): 
                        board[r][c] = 'S' 
                        q.append((r, c)) 

        print(board)

        while q: 
            r, c = q.popleft() 

            for dr, dc in [(1, 0), (-1, 0), (0, 1), (0, -1)]: 
                nr, nc = r + dr, c + dc 

                if (0 <= nr < rows and 
                    0 <= nc < cols and  
                    board[nr][nc] == 'O'): 
                        board[nr][nc] = 'S' 
                        q.append((nr, nc)) 

        for r in range(rows): 
            for c in range(cols): 
                if board[r][c] == 'S': 
                    board[r][c] = 'O' 
                elif board[r][c] == 'O': 
                    board[r][c] = 'X'

File: Data Structures & Algorithms/search-for-word/submission-15.js
class Solution {
    /**
     * @param {character[][]} board
     * @param {string} word
     * @return {boolean}
     */
    exist(board, word) {   
        const rows = board.length 
        const cols = board[0].length
         
        const dfs = (i, r, c, visited) => { 
            if (i === word.length) return true 

            const pos = `${r},${c}` 
            if (!( 
                r >= 0 && 
                c >= 0 && 
                r < rows && 
                c < cols &&
                board[r][c] === word[i] && 
                !visited.has(pos)
            )) return false 

            visited.add(pos) 
            const res = dfs(i + 1, r + 1, c, visited) || 
                        dfs(i + 1, r, c + 1, visited) || 
                        dfs(i + 1, r - 1, c, visited) || 
                        dfs(i + 1, r, c - 1, visited) 
            visited.delete(pos) 

            return res
        } 

        for (let r = 0; r < rows; r ++) {
            for (let c = 0; c < cols; c ++) {
                if (board[r][c] === word[0]) {
                    const res = dfs(0, r, c, new Set()) 
                    if (res) return true
                }
            }
        } 

        return false
    }
}
